### ISSUE TITLE
Check for missing handling of extended state (FPU, SSE, AVX, etc.)

### ISSUE THREAD

**Issue by HW42** [oscarlab/graphene#627]
PR #626 fixes leaking of information in "extended" registers.

We need to check where else we are missing handling of extended state:

 - Should we reset the extended state to a known state on enclave entry?
 - Before exiting for an ocall we only call `fxsave`. The Intel SDK uses `xsave` here. OTOH AFAIU only `MXCSR` and `FCW` are callee saved.
 - The exception handler needs to save extended state.

Anything else?

**Comment by dimakuv** [comment]
> The exception handler needs to save extended state.
Is #397 a step in the right direction?

**Comment by dimakuv** [comment]
> Should we reset the extended state to a known state on enclave entry?

Yes, we can re-use the clean XSAVE area state introduced in #626. This should go before `callq handle_ecall`.

**Comment by yamahata** [comment]
To prevent information leak, those three above would cover it.

I dumped what I have as #632.
_DkHandleExternalEvent can call exception handler which requires the context of fp registers. 
Possibly we can simply zero the saving area except MXCSR and FCW instead of xsave/fxsave.
For correct ABI for user application, fp context on entering LibOS should be passed.
I don't know any application that uses fp context on entering kernel by signal handler, though.
In practice, it wouldn't matter.

**Comment by HW42** [comment]
Ok, so AFAICS we have:

 - Clear before EEXIT is implemented in PR #626 and as part of PR #632.
 - PR #632 covers ocalls and exception handling, I expect that @yamahata will provide separated PRs soon-ish.
 - I can provide a simple follow up to PR #626 to clear before ecall handler.

@yamahata: What do you think? How show we resolve PR #626 vs your changes?

**Comment by yamahata** [comment]
@HW42 sounds good.

**Comment by dimakuv** [comment]
Btw, we are planning to merge 626 soon.

**Comment by mkow** [comment]
FYI: #626 was just merged.

**Comment by HW42** [comment]
> - I can provide a simple follow up to PR #626 to clear before ecall handler.

This is a bit more complex than I thought. The problem is that we didn't read the value of XFRM yet before calling the ecall handler. I see 3 options:

 1. Only do fxrstor in early assembly and delay clearing the rest (i.e. features not covered by fxrstor) to `enclave_init`. In this case I would wait for the XSAVE relared parts of PR 632 since I also need to call xrstor from C.
 2. Get the XFRM value in assembly. This should not be that complex and would make the code which clears before EEXIT cleaner (since we no longer depend on the compiler to not use AVX & co. before we read XFRM).

Is there any hardware which supports SGX but not XSAVE, and do we want to support this? If not there's an easy third option:

 3. Always set XFRM to something > 3. Then the enclave is only entered if XSAVE is available. This would make all the related code easier.

**Comment by mkow** [comment]
2. seems to be the best option to me, it doesn't sound too complex and doesn't require any additional assumptions.

**Comment by dimakuv** [comment]
Section 42.7.2.1 of the Intel Software Developer Manual states this:

> On processors not supporting XSAVE, software should initialize XFRM to 0x3. On processors supporting XSAVE, software should initialize XFRM to be a subset of XCR0 that would be present at the time of enclave execution. Because bits 0 and 1 of XFRM must always be set, the use of Intel SGX requires that SSE be enabled (CR4.OSFXSR = 1).

So it is possible to have a CPU which supports SGX but not XSAVE.

So Option 2 is the most reasonable.

**Comment by dimakuv** [comment]
@HW42 @yamahata Reviving this old thread. The missing extended-state handling bites us again!

While debugging a mysterious SIGFPE fault on this instruction during an OpenVINO run:
```
addss  0x140e08(%rip),%xmm0  # from Glibc 2.29 function <round_and_return+824>
```
it results in a SIGFPE with the "Numeric underflow (#U)" SSE hardware exception. The exception is raised because the UM bit is not set in the MXCSR register (bit 11, meaning "Mask bit clear" -> CPU raises the #XM/#XF exception with bit #UE set in MXCSR).

The root cause is interesting. This SIGFPE only happens on the *second* enclave thread and not on the *main* enclave thread. Note how we currently do not sanitize the XSAVE area (which includes MXCSR) on enclave-thread entry (what you call "enclave entry"). So the main thread contains the default value of MXCSR = 0x1f80 (all mask bits are set), but after `shim_do_clone()` the second enclave thread has the value of MXCSR = 0x0 (all mask bits are not set). Therefore, the main thread does not fail on the above instruction, but the second thread does.

This can be reproduced by this example:
```
#include <xmmintrin.h>
#include <pthread.h>
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>

void* child_print(void* arg) {
//    _mm_setcsr(0x1f80);  /* uncomment to get rid of SIGFPE in Graphene-SGX */
    float res = strtof("1.000000013351432e-10", NULL);
    printf("child: %f\n", res);
    return NULL;
}

int main(int argc, char** argv) {
    float res = strtof("1.000000013351432e-10", NULL);
    printf("parent: %f\n", res);

    pthread_t thread;
    pthread_create(&thread, NULL, child_print, NULL);
    pthread_join(thread, NULL);
    return 0;
}
```

Graphene-SGX will fail with something like:
```
[     2] arithmetic fault at 0x048a1cd8
[     2] SIGFPE handled
```

This is because "1.000000013351432e-10" is a denormalized FP value, so CPU wants to trigger the exception, and since `MXCSR[bit 11] = 0`, it actually raises the exception.

As soon as we update MXCSR to a value with bit equal to 1 (e.g. `MXCSR = 0x1f80`), the CPU masks the exception, and Graphene-SGX continues as expected.

I will submit a PR now with my quick fix. **But looking at this thread, @HW42 thinks that this is non-trivial.** I am not sure I understood why.

**Comment by dimakuv** [comment]
@HW42, Ok, I understood why adding the FXRSTOR/XRSTOR logic to the enclave-thread entry is hard. This is because we don't have a simple way to figure out whether XRSTOR is supported on this machine at this point (this would require calculating the correct address somewhere in SECS/TCS in assembly).

To fix this particular bug, at least FXRSTOR is enough. So I implemented your option 1 currently in the submitted PR.

**Comment by dimakuv** [comment]
This was fixed long ago with #1064.
